### PR TITLE
Iter2 ross part2

### PR DESCRIPTION
**Pull request by perryr16**
added comments to many methods

deliverable methods:
highest_scoring_visitor
highest_scoring_home_team
lowest_scoring_visitor
lowest_scoring_home_team

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -21,26 +21,28 @@ def self.count_of_games_by_season
...
   def self.games_per(csv_header)
+    #returns number of games in a given (:season) or (:away_team_id)...
     group_by_header = @@all.group_by { |game| game.send(csv_header) }
     group_by_header.values.map{ |games| games.length}
   end
 
 
-  def self.goals_per(csv_header, goals)
+  def self.goals_per(csv_header, hoa_goals)
+    #returns number of goals (:home_goals) or (:away_goals) in a given csv_header
```
Comment (perryr16) [inline lib/game.rb:38]: general description of what the method does and what kind of arguments need to be passed in

File: lib/game.rb [hunk 2]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
       (goals.to_f / length_array[index].to_f ).round(2)
     end
   end
-#one of THE METHODS ross
+#deliverable
   def self.average_goals_by_season
-    goals = home_away_goals_per(:season)
-    season_length = games_per(:season)
-    average_goals_per_season = average_goals(goals, season_length)
+    avg_goals_per_season = average_goals(total_goals_per(:season), games_per(:season))
```
Comment (perryr16) [inline lib/game.rb:56]: removed many variables, instead just showing average goals per season

File: lib/game.rb [hunk 3]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
-    games_length = games_per(hoa_team_id)
-    avg_goals = average_goals(goals, games_length)
-    team_id_goals = Hash[team_ids.zip(avg_goals)]
-    team_id_goals.send(max_min_by){ |team_id, hoa_goals| hoa_goals}.first
+
+#deliverable (needs to access teams.csv via stat_tracker)
+  def self.highest_scoring_visitor_team_id
+    avg_away_goals = average_goals(goals_per(:away_team_id, :away_goals), games_per(:away_team_id))
+    away_team_ids = @@all.map { |game| game.away_team_id }.uniq
+    # create hash of {team_ids => average goals}
```
Comment (perryr16) [inline lib/game.rb:66]: preveiw of what the new hash will look like

File: lib/game.rb [hunk 4]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+    away_ids_n_goals = Hash[away_team_ids.zip(avg_away_goals)]
+    away_ids_n_goals.max_by{ |team_id, away_goals| away_goals}.first
+  end
+#deliverable (needs to access teams.csv via stat_tracker)
+  def self.highest_scoring_home_team_id
```
Comment (perryr16) [inline lib/game.rb:71]: these four methods are all VERY similar, one or two variables changed in each

File: lib/game.rb [hunk 5]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+#deliverable (needs to access teams.csv via stat_tracker)
+  def self.lowest_scoring_home_team_id
+    avg_home_goals = average_goals(goals_per(:home_team_id, :home_goals), games_per(:home_team_id))
+    home_team_ids = @@all.map { |game| game.home_team_id }.uniq
+    # create hash of {team_ids => average goals}
+    home_ids_n_goals = Hash[home_team_ids.zip(avg_home_goals)]
+    home_ids_n_goals.min_by{ |team_id, home_goals| home_goals}.first
+  end
+
+  def self.games_by_season(team_id)
```
Comment (perryr16) [inline lib/game.rb:95]: helper method for best season

File: lib/game.rb [hunk 6]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+    end
+    season_wins
+  end
+
+  def self.percent_by_season(team_id)
+    #this each_with_index can perform mult/div on arrays of the same size
+    # [1,2,3] * [11,12,13] = [11,24,39]
+    num_games = games_by_season(team_id).values
+    percent_per_season = wins_by_season(team_id).values.each_with_index.map do |wins, index|
+      ((wins.to_f / num_games[index].to_f)*100).to_i
```
Comment (perryr16) [inline lib/game.rb:126]: this is my best attempt to divide two arrays of the same size. let me know if anyone has a better way

File: lib/game.rb [hunk 7]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+    end
+
+    Hash[games_by_season(team_id).keys.zip(percent_per_season)]
+  end
+
+#deliverable
+  def self.best_season(team_id)
+    #return season with highest winning percentage
+    best_season = percent_by_season(team_id).max_by { |season, percent| percent}
+    "In the #{best_season[0]} season Team #{team_id} won #{best_season[1]}% of games"
```
Comment (perryr16) [inline lib/game.rb:136]: I improvised on the return string

File: lib/stat_tracker.rb [hunk 8]
```diff
@@ -60,35 +60,44 @@ def percentage_ties
   def count_of_games_by_season
     Game.count_of_games_by_season
   end
-  #ross
```
Comment (perryr16) [inline lib/stat_tracker.rb:63]: no one needs to see my name

File: test/fixtures/games_20.csv [hunk 9]
```diff
@@ -5,7 +5,7 @@ game_id,season,type,date_time,away_team_id,home_team_id,away_goals,home_goals,ve
 2016030173,20162017,Postseason,4/18/17,24,20,3,2,BMO Field,/api/v1/venues/null
 2014030412,20142015,Postseason,6/6/15,16,14,3,2,Audi Field,/api/v1/venues/null
 2014030413,20142015,Postseason,6/9/15,14,16,3,2,Gillette Stadium,/api/v1/venues/null
-2014030131,20142015,Postseason,4/16/15,5,3,1,2,BBVA Stadium,/api/v1/venues/null
+2014030131,20142015x,Postseason,4/16/15,5,3,1,2,BBVA Stadium,/api/v1/venues/null
```
Comment (perryr16) [inline test/fixtures/games_20.csv:8]: late night attempt to prove sanity. it failed

File: test/game_test.rb [hunk 10]
```diff
@@ -103,33 +102,83 @@ def test_it_can_calculate_highest_total_score
...
     assert_equal 3, @game.lowest_total_score
   end
-
+#deliverable
   def test_it_can_count_games_by_season
     assert_equal ({20122013=>2, 20162017=>5, 20142015=>6, 20132014=>4, 20152016=>2, 20172018=>1}), Game.count_of_games_by_season
   end
+#deliverable
+  def test_highest_scoring_visitor_team_id
+    Game.all.stubs(:map).returns([1, 2, 3, 4])#team_ids
```
Comment (perryr16) [inline test/game_test.rb:111]: I should avoid stubbing :map

File: test/game_test.rb [hunk 11]
```diff
@@ -103,33 +102,83 @@ def test_it_can_calculate_highest_total_score
...
+    assert_equal "In the 20142015 season Team 3 won 100% of games", Game.best_season(3)
+    assert_equal "In the 20122013 season Team 6 won 100% of games", Game.best_season(6)
+    assert_equal "In the 20162017 season Team 20 won 0% of games", Game.best_season(20)
+    stub_val = {
+                20122013 => 25,
+                20132014 => 66,
+                20142015 => 44,
+                20152016 => 35,
+                }
+    Game.stubs(:percent_by_season).returns(stub_val)
```
Comment (perryr16) [inline test/game_test.rb:167]: I think this is an effective stub

**Review by perryr16 — COMMENTED** [review]
ross's comments

File: lib/game.rb (hunk 1 above)
Comment (stellakunzang) [reply lib/game.rb:38]: thank you, this is helpful!

File: test/game_test.rb (hunk 10 above)
Comment (stellakunzang) [reply test/game_test.rb:111]: correct

File: test/fixtures/games_20.csv (hunk 9 above)
Comment (perryr16) [reply test/fixtures/games_20.csv:8]: Gif game is out of control. 10 out of 10

File: test/game_test.rb (hunk 10 above)
Comment (Rostammahabadi) [reply test/game_test.rb:111]: agreed

File: test/game_test.rb (hunk 11 above)
Comment (Rostammahabadi) [reply test/game_test.rb:167]: agreed

**Comment by Rostammahabadi** [comment]
Looks good

File: lib/game.rb [hunk 12]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
-#FOUR of THE METHODS ross
-  def self.nth_scoring_team_id(max_min_by, hoa_team_id, hoa_goals)
-    team_ids = @@all.map { |game| game.send(hoa_team_id) }.uniq
-    goals = goals_per(hoa_team_id, hoa_goals)
-    games_length = games_per(hoa_team_id)
-    avg_goals = average_goals(goals, games_length)
-    team_id_goals = Hash[team_ids.zip(avg_goals)]
-    team_id_goals.send(max_min_by){ |team_id, hoa_goals| hoa_goals}.first
+
+#deliverable (needs to access teams.csv via stat_tracker)
```
Comment (Rostammahabadi) [inline lib/game.rb:62]: These comments are going to make it 100X easier to navigate so thanks for this!

File: lib/game.rb [hunk 13]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+  def self.games_by_season(team_id)
 
+    #accumulator hash
+    games_by_season = Hash.new { |hash, key| hash[key] = 0 }
+    @@all.each do |game|
+      #returns true if a team was in a given game
+      team_played_in_game = game.away_team_id == team_id || game.home_team_id == team_id
+      games_by_season[game.season] += 1 if team_played_in_game
+    end
+    games_by_season
```
Comment (Rostammahabadi) [inline lib/game.rb:104]: Is the return value for games by season an int?

File: lib/game.rb [hunk 14]
```diff
@@ -49,24 +51,97 @@ def self.average_goals(sum_array, length_array)
...
+  end
```
Comment (Rostammahabadi) [inline lib/game.rb:105]: This is a great way to organize by season for a team

File: lib/game.rb [hunk 15]
```diff
@@ -99,4 +174,10 @@ def highest_total_score
   def lowest_total_score
     @@all.map { |game| game.away_goals + game.home_goals}.min
   end
+
+  def win?(team_id)
+    away_win = team_id == @away_team_id && @away_goals > @home_goals
+    home_win =  team_id == @home_team_id && @home_goals > @away_goals
+    away_win || home_win
+  end
```
Comment (Rostammahabadi) [inline lib/game.rb:182]: This will come in handy when refactoring

**Review by Rostammahabadi — APPROVED** [review]
Everything looks good

File: lib/game.rb (hunk 1 above)
Comment (foleymichelle9) [reply lib/game.rb:38]: Agreed Thank you!

**Pull request merged**